### PR TITLE
opt: remove tuple lables from subquery output

### LINKED ISSUE #28240: opt: top-level tuple labels shouldn't get returned across subqueries
In the following query, the optimizer attaches identical unknown (`?column?`) tuple labels to the tuple returned by the inner subquery. This confuses the execution layer, which may try to serialize that tuple with its labels to distribute. Upon deserializing the tuple, evaluation will fail because the tuple will have duplicate columns.

To behave like the heuristic planner in this regard, the optimizer should strip labels from the tuple returned by this kind of subquery.

```
SELECT (1, 2, 3) IN (SELECT 1, 2, 3)
```

The optimizer's output, post subquery replacement:

```
ANY (((1, 2, 3) AS "?column?", "?column?", "?column?"),): found duplicate tuple label: "?column?"
```

### PR DESCRIPTION
**Pull request by rytaft**
Tuple types include an optional label for each column in the
tuple. Previously, the optimizer was populating these labels for
the output of subqueries. However, this was causing a problem in
the execution engine for queries such as:
```
SELECT (1, 2, 3) IN (SELECT 1, 2, 3)
```
This query causes an error:
```
ANY (((1, 2, 3) AS "?column?", "?column?", "?column?"),): found
duplicate tuple label: "?column?"
```
This commit removes the labels from the type.

Fixes #28240

cc @jordanlewis 

Release note: None

### CODE REVIEW

**Comment by rytaft** [comment]
If we want to preserve the labels in the type but remove them from the tuple itself, another (less clean) option is to remove the tuple labels here and a few other places: https://github.com/cockroachdb/cockroach/blob/8044dea9c5c210c3ddb238a7edd446a93013e2e2/pkg/sql/opt/exec/execbuilder/scalar_builder.go#L387

I think the only reason to preserve the labels in the type is for explain output.

**Review by jordanlewis — COMMENTED** [review]
Confirmed that this fixes the issue on my branch. Fix also :lgtm: - seems straightforward!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28258#-:-LJ0XTlY-xH40ubtxNXK:btkhfdr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by rytaft** [comment]
Thanks! I'll wait for confirmation from @knz before merging, since I vaguely remember that this was a concern for explain output. But the alternative approach of removing the labels in the execbuilder is much messier, so I'd rather not go that route unless necessary (unless you see a clean way to do it, @RaduBerinde).

**Review by RaduBerinde — COMMENTED** [review]
With this change, does something like this work:
```
radu=> SELECT (x).n FROM (SELECT information_schema._pg_expandarray(ARRAY[1,2,3])) AS t(x);
 n 
---
 1
 2
 3
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28258#-:-LJ1SWD76Vv04eh8MkiD:b-7ybb96)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Review by RaduBerinde — COMMENTED** [review]
Ah wait, that's not even a subquery. I was thinking of something more along the lines of this, but it doesn't work in PG either:

```
radu=> SELECT (SELECT information_schema._pg_expandarray(ARRAY[1])).n;
ERROR:  could not identify column "n" in record data type
LINE 1: SELECT (SELECT information_schema._pg_expandarray(ARRAY[1]))...
               ^
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28258#-:-LJ1_FRx8uQN3fyYP9DW:b-rmf6o1)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Review by rytaft — COMMENTED** [review]
Yea, the first query works. The second one causes a syntax error.

I'm pretty confident that this change only affects the `TYPES` explain output. @knz, is that a problem? Ok to merge this?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28258#-:-LJEcjx06YXClM8tTxC8:b-e9ap1r)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Review by knz — APPROVED** [review]
:lgtm: yes this is good, thanks for catching this

(for context, indeed also in postgres the tuple type generated for a SQ result schema never has labels)

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28258#-:-LJIkUqP8tHuz0G4GFs7:bxoa5ec)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

**Review by rytaft — COMMENTED** [review]
Thanks! 

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28258#-:-LJJlQ2y8T8EyO_-BABR:b-hlp5ya)* status: :shipit: complete! 2 of 0 LGTMs obtained

**Comment by RaduBerinde** [comment]
I think we hit the same flaky test, which I'm fixing in #28329.

bors r+

**Comment by jordanlewis** [comment]
Flake...

bors r+

**Pull request merged**
